crates/flox-rust-sdk/src/models/search.rs

```diff
@@ -0,0 +1,261 @@
+}
+
+/// A package source
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct RegistryInput {
+    /// The flake containing packages
+    pub from: FlakeRef,
+    /// An optional attr path to restrict the search to
+    pub subtrees: Option<Vec<String>>,
+    /// Which stabilities should be included in the search results
+    pub stabilities: Option<Vec<Stability>>,
+}
+
+/// Which packages should be allowed in search results
+#[derive(Debug, Clone, Serialize, Deserialize, Default)]
+pub struct AllowOpts {
+    /// Whether packages with unfree licenses should be included
+    pub unfree: bool,
+    /// Whether packages that are marked "broken" should be included
+    pub broken: bool,
+    /// A whitelist of package licenses
+    pub licenses: Option<Vec<String>>,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, Default)]
+#[serde(rename_all = "camelCase")]
+pub struct SemverOpts {
+    pub prefer_pre_releases: bool,
+}
+
+/// A set of options for defining a search query.
+///
+/// The search options aren't mutually exclusive. For instance, the query
+/// `hello@>=2` will populate the `match` field with `hello` and the `semver`
+/// field with `>=2`. The `match` field specifically searches the `name`, `pname`,
+/// and `description` fields.
+///
+/// The result of the query will be the logical AND of all provided parameters.
+#[derive(Debug, Clone, Serialize, Deserialize, Default)]
+pub struct Query {
+    /// Match against the full name of the package e.g. `<pname>-<version>`
+    pub name: Option<String>,
+    /// Match against the `pname` of the package
+    pub pname: Option<String>,
+    /// Match against the explicit version number of the package
+    pub version: Option<String>,
+    /// Match against a semver specifier
+    pub semver: Option<String>,
+    /// Match against a regular expression
+    pub r#match: Option<String>,
+}
+
+impl FromStr for Query {
+    type Err = SearchError;
+
+    // This can't actually error, but the trait requires an error type
+    fn from_str(search_term: &str) -> Result<Self, Self::Err> {
+        let q = match search_term.find('@') {
+            Some(idx) => {
+                // If we get a search term ending in '@' it most likely means the
+                // user didn't quote a search term that included a '>' character.
+                if idx >= search_term.len() - 1 {
+                    return Err(SearchError::SearchTerm(search_term.into()));
+                }
+                // Splitting at `idx` would include the `@`
+                let package = String::from(&search_term[..idx]);
+                let semver = String::from(&search_term[idx + 1..]);
+                Query {
+                    semver: Some(semver),
+                    r#match: Some(package),
+                    ..Query::default()
+                }
+            },
+            None => Query {
+                r#match: Some(search_term.into()),
+                ..Query::default()
+            },
+        };
```

Comment:
```suggestion
        let q = match search_term.split_once('@') {
            Some((package, semver) => {
                // If we get a search term ending in '@' it most likely means the
                // user didn't quote a search term that included a '>' character.
                if semver.is_empty() {
                    return Err(SearchError::SearchTerm(search_term.into()));
                }
                Query {
                    semver: Some(semver.to_string()),
                    r#match: Some(package.to_string()),
                    ..Query::default()
                }
            },
            None => Query {
                r#match: Some(search_term.to_string()),
                ..Query::default()
            },
        };
```

To me using split_once and is_empty is a clearer story than messing with string indexes.

Reply:
Nice, I was looking for a method that wouldn't include the delimiter and must have missed that.

Reply:
Fixed